api/v1alpha1/contour_types.go

```diff
@@ -266,6 +266,11 @@ type ProviderLoadBalancerParameters struct {
 	// +unionDiscriminator
 	// +kubebuilder:default=AWS
 	Type LoadBalancerProviderType `json:"type,omitempty"`
+
+	// loadBalancerIP contains IP for LoadBalancer service type, optional
+	// +kubebuilder:validation:Format=ipv4
```

Comment:
should we allow ipv6 ips as well?

Reply:
I was thinking about it, but I am not sure how to handle this. Problem is that of course format can only be specified one, so e.g. I can not do `Format=ipv4;ipv6`. So I was thinking, either we move everything to `validation.go`, we have separate field (e.g. `LoadBalancerIPv6`), or maybe simplest way would be to just use same spec as it is implemented into * corev1.ServiceSpec (`json:"loadBalancerIP,omitempty" protobuf:"bytes,8,opt,name=loadBalancerIP"`).

What do you think, maybe last option would make the most sense as it follows same structure as corev1?

Reply:
Maybe we can leave it as IPv4 for now and if there is some requirement for IPv6, we can add support later?

Reply:
i would be on the side of letting ipv6 addresses be specified to start (probably via adding to the validation code we have today, probably simple enough to use `net.ParseIP`), though leaving things restricted and opening it up when needed make sense

I would like to get other @projectcontour/maintainers thoughts on this as well

Reply:
I would expect `LoadBalancerIP` to be a field of `LoadBalancerStrategy` or `ProviderLoadBalancerParameters`. @BostjanBozic a few questions:
- Currently `LoadBalancerStrategy` supports LB scope. Must a loadbalancer have an external scope to specify an IP?
- Do the major cloud providers support specifying a LB IP? From the Service spec: "This feature depends on whether the underlying cloud-provider supports specifying the loadBalancerIP when a load balancer is created. This field will be ignored if the cloud-provider does not support the feature." If this feature is not widely supported, it should be part of `ProviderLoadBalancerParameters`.

Reply:
@danehans Cheers for feedback. I put that under `ProviderLoadBalancerParameters`, as I was thinking this is parameter of LB itself, just for naming semantics. If preferred I can move it to `LoadBalancerStrategy` as well.
Regarding questions:
* As far as I am aware, you can also use this when `externalTrafficPolicy: Local`, but you need to provide IP in correct subnet
* I did check AWS, Azure and GCP. For Azure and GCP they support explicitly configuring `loadBalancerIP` (with Azure you might have to configure annotation in case public IP resource is configured in another resource group - https://github.com/projectcontour/contour-operator/issues/327). For AWS, you have to provide annotation `service.beta.kubernetes.io/aws-load-balancer-eip-allocations`. Good thing is if cloud provider does not support setting `loadBalancerIP`, this field will be ignored.

Reply:
`ProviderLoadBalancerParameters` is meant for passing provider-specific LB config and why I'm trying to understand what cloud providers support specifying a LB IP. If the major cloud providers support the ability to specify an IP, then the field should be in `LoadBalancerStrategy`. 

> As far as I am aware, you can also use this when externalTrafficPolicy: Local, but you need to provide IP in correct subnet

I'm referring to [this scope](https://github.com/projectcontour/contour-operator/blob/release-1.15/api/v1alpha1/contour_types.go#L253-L257), which is used to specify an internal or external LB. All major cloud providers support specifying an internal load-balancer, hence why `scope` is part of `LoadBalancerStrategy` and not `ProviderLoadBalancerParameters`. I'm trying to understand how specifying `loadBalancerIP` will work with `scope: internal`. Do cloud providers allow an internal LB scope when `loadBalancerIP` is set? If so, so any limitations exist? For example, cloud provider foo allows `loadBalancerIP` with internally-scoped LB but the address must be from "10.x.x.x/8".

Reply:
It is only Azure and GCP that are currently supporting providing exact IP. In AWS's case, binding existing LB to a service requires annotating the service. So basically following is the statue:
* Azure: you can specify `loadBalancerIP` parameter, in case of Public IP resource being in another resource group, you have to annotate service with `service.beta.kubernetes.io/azure-load-balancer-resource-group: myResourceGroup`
* GCP: you can specify `loadBalancerIP` parameter, I think nothing else is required
* AWS: specifying `loadBalancerIP` parameter has no effect, service needs annotation (`service.beta.kubernetes.io/aws-load-balancer-eip-allocations: eipalloc-<xxxxxxxxxxxxxxxxx>`)

Regarding the scope, I was only able to test Azure. Using internal scope following results occur:
* `loadBalancerIP` parameter can be configured
* `loadBalancerIP` needs to be in specific range
* when creating Public IP on Azure, this will never be in range required above, hence preconfiguring `loadBalancerIP` parameter for `Internal` scope makes no sense, at least for Azure

Reply:
I have used GCP in this way - allocate a regional static IP address, and that set that IP address to the `loadBalancerIP` in the service, and it works. We do it this way because we have all our networking created in terraform - so we let terraform create the IP addresses, and then also create the DNS records that depend on those IP addresses. Then when we run our k8s deployment, we get the IP from terraform and plug it into our k8s deploy scripts.